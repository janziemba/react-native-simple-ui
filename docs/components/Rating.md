---
id: rating
title: Rating
---
## Props
- `emptyIcon` (_object_, **required**)
    - [`Icon` props](Icon.md#props)
- `fullIcon` (_object_, **required**)
    - [`Icon` props](Icon.md#props)
- `halfIcon` (_object_, **required**)
    - [`Icon` props](Icon.md#props)
- `limit` (_number_, default 5)
- `onPress` (_function_, default null)
- `rating` (_number_, **required**)

## Examples

```jsx

```
