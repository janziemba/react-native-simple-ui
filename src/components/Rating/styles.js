// @flow

import { StyleSheet } from 'react-native';

export default (): StyleSheet.Styles => ({
    alignItems: 'center',
    flexDirection: 'row',
});
