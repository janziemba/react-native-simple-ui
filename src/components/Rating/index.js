// @flow

import React, { PureComponent } from 'react';
import type { Node } from 'react';
import { StyleSheet, TouchableOpacity, View } from 'react-native';

import withTheme from '../../themes/withTheme';
import Icon from '../Icon';
import type { Props as IconProps } from '../Icon';
import styles from './styles';

type Props = {
    emptyIcon: IconProps,
    fullIcon: IconProps,
    halfIcon: IconProps,
    limit?: number,
    onPress?: Function,
    rating: number,
    styles: StyleSheet.Styles,
};

class Rating extends PureComponent<Props> {
    static defaultProps = {
        limit: 5,
        onPress: null,
    };
    constructor(props: Props): void {
        super(props);

        (this: any).renderEmptyIcons = this.renderEmptyIcons.bind(this);
        (this: any).renderFullIcons = this.renderFullIcons.bind(this);
        (this: any).renderHalfIcon = this.renderHalfIcon.bind(this);
    }
    renderEmptyIcons(): Array<Node> {
        const { emptyIcon, limit, rating } = this.props;

        const nearestHalf: number = Math.round(rating * 2) / 2;
        const activeCount: number = nearestHalf % 1 === 0
            && nearestHalf > rating ? Math.ceil(rating) : Math.floor(rating);
        const halfShouldShow: boolean = nearestHalf % 1 > 0;
        const count: number = (limit || 5) - activeCount - (halfShouldShow ? 1 : 0);
        const icons: Array<Node> = [];

        for (let i = 0; i < count; i += 1) {
            icons.push(<Icon
                {...emptyIcon}
                key={`emptyIcon_${i}`}
            />);
        }

        return icons;
    }
    renderFullIcons(): Array<Node> {
        const { fullIcon, rating } = this.props;

        const nearestHalf: number = Math.round(rating * 2) / 2;
        const count: number = nearestHalf % 1 === 0
            && nearestHalf > rating ? Math.ceil(rating) : Math.floor(rating);
        const icons: Array<Node> = [];

        for (let i = 0; i < count; i += 1) {
            icons.push(<Icon
                {...fullIcon}
                key={`fullIcon_${i}`}
            />);
        }

        return icons;
    }
    renderHalfIcon(): Node {
        const { halfIcon, rating } = this.props;

        const nearestHalf: number = Math.round(rating * 2) / 2;
        const shouldShow: boolean = nearestHalf % 1 > 0;

        if (!shouldShow) {
            return null;
        }

        return (
            <Icon {...halfIcon} />
        );
    }
    render(): Node {
        const { onPress, styles } = this.props;

        return (
            <TouchableOpacity
                onPress={onPress}
            >
                <View style={styles}>
                    {this.renderFullIcons()}
                    {this.renderHalfIcon()}
                    {this.renderEmptyIcons()}
                </View>
            </TouchableOpacity>
        );
    }
}

Rating = withTheme(styles, 'Rating')(Rating);

export default Rating;
