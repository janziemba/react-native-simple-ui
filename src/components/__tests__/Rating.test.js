import React from 'react';
import renderer from 'react-test-renderer';

import Rating from '../Rating';

it('renders correctly', () => {
    const tree = renderer.create(
        <Rating
            emptyIcon={{
                name: 'star-border',
            }}
            fullIcon={{
                name: 'star',
            }}
            halfIcon={{
                name: 'star-half',
            }}
            rating={3.2}
        />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
