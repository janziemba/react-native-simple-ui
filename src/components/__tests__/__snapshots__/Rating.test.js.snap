// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": null,
            "fontSize": 24,
          },
          undefined,
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": null,
            "fontSize": 24,
          },
          undefined,
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": null,
            "fontSize": 24,
          },
          undefined,
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": null,
            "fontSize": 24,
          },
          undefined,
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": null,
            "fontSize": 24,
          },
          undefined,
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
